connector/connector_ldap.go

```diff
@@ -217,11 +238,74 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+type LDAPPool struct {
+	m          sync.Mutex
+	conns      map[*ldap.Conn]*ldap.Conn
+	initConn   int
+	maxConn    int
+	serverHost string
+	serverPort uint16
+	useTLS     bool
+	useSSL     bool
+	tlsConfig  *tls.Config
+}
+
+func (p *LDAPPool) fillPool() error {
+	if p.conns == nil {
+		p.conns = make(map[*ldap.Conn]*ldap.Conn)
+	}
+	for i := len(p.conns); i < p.initConn; i++ {
+		ldapConn, err := p.ldapConnect()
+		if err != nil {
+			return err
+		}
+		p.conns[ldapConn] = ldapConn
+	}
+	return nil
+}
+
+func (p *LDAPPool) Get() (*ldap.Conn, error) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	if p.conns == nil {
+		err := p.fillPool()
+		if err != nil {
+			return nil, err
+		}
+	}
+	for conn := range p.conns {
+		delete(p.conns, conn)
+		return conn, nil
+	}
+	// the pool is empty
+	conn, err := p.ldapConnect()
+	return conn, err
+}
+
+func (p *LDAPPool) Put(c *ldap.Conn) {
+	p.m.Lock()
+	defer p.m.Unlock()
+
+	if len(p.conns)+1 > p.maxConn {
+		c.Close()
+	} else {
+		// drop to anonymous bind
+		err := c.Bind("", "")
```

Comment:
Like the comment above, this bind should be performed outside of the mutex lock. The mutex should be used to guard the map and the counter, don't make remote calls while holding it.